crates/typst/src/export/svg.rs

```diff
@@ -0,0 +1,626 @@
+use std::{
+    collections::HashMap,
+    fmt::{Display, Write},
+    hash::Hash,
+    io::Read,
+};
+
+use base64::Engine;
+use ttf_parser::{GlyphId, OutlineBuilder};
+use xmlwriter::XmlWriter;
+
+use crate::{
+    doc::{Document, Frame, FrameItem, Glyph, GroupItem, TextItem},
+    font::Font,
+    geom::{Abs, Axes, Geometry, LineCap, LineJoin, PathItem, Ratio, Shape, Transform},
+    image::{ImageFormat, RasterFormat, VectorFormat},
+    util::hash128,
+};
+use crate::{geom::Paint::Solid, image::Image};
+
+/// [`RenderHash`] is a hash value for a rendered glyph or clip path.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+struct RenderHash(u128);
+
+/// Convert a [`u128`] into a [`RenderHash`].
+impl From<u128> for RenderHash {
+    fn from(value: u128) -> Self {
+        Self(value)
+    }
+}
+
+/// Export a document into a SVG file.
+#[tracing::instrument(skip_all)]
+pub fn svg(doc: &Document) -> String {
+    let mut renderer = SVGRenderer::new();
+    let max_page_width = doc
+        .pages
+        .iter()
+        .map(|page| page.size().x)
+        .max_by(|a, b| a.partial_cmp(b).unwrap())
+        .unwrap_or(Abs::zero());
+    let total_page_height = doc.pages.iter().map(|page| page.size().y).sum::<Abs>();
+    let doc_size = Axes { x: max_page_width, y: total_page_height };
+    renderer.header(doc_size);
+    let mut y_offset = Abs::zero();
+    for page in &doc.pages {
+        renderer.render_frame(page, Transform::translate(Abs::zero(), y_offset));
+        y_offset += page.size().y;
+    }
+    renderer.finalize()
+}
+
+/// Export a frame into a SVG file.
+#[tracing::instrument(skip_all)]
+#[comemo::memoize]
```

Comment:
Is it possible to memorize this function? I expect changes typically only affect one page.

Reply:
Should be okay. We don't currently do it for `render` although it's similar, but (a) we could and (b) if we decide not to because of high memory usage, the situation might be different here because SVGs are smaller. The return value will be cloned so a cheap-to-clone type would be preferrable. Typically, we use `EcoString` for this, but `XmlWriter` uses `Vec<u8>` internally and the conversion to `EcoString` would mean a copy, so maybe `Arc<String>` is better here.

Reply:
> The return value will be cloned so a cheap-to-clone type would be preferrable. Typically, we use `EcoString` for this, but `XmlWriter` uses `Vec<u8>` internally and the conversion to `EcoString` would mean a copy, so maybe `Arc<String>` is better here. Not sure which I prefer to be honest.

`Arc<str>` might be even better.

Reply:
`Arc<str>` makes sense to me, given that the result string is unlikely needed to be changed

Reply:
I will remove this because I find that the hash actually changes even if the corresponding frame doesn't contains visible differences. 

```
hash for frame 0: 166703367601710688816886167189174114022
hash for frame 1: 98367643144455113841213878673356476442
hash for frame 2: 32996071146373462600790487505688723979
hash for frame 3: 306199407462341050674753856980979556156
hash for frame 4: 230856579288412513524522021777448394661
hash for frame 5: 230494102445966875144355170909939471589
hash for frame 6: 335696108594705439680628222702518971520
hash for frame 7: 88611471193250368192454119476743850054
hash for frame 8: 85320809752456081553206451887523253111
hash for frame 9: 85089647697418920685636420564672793130
hash for frame 10: 277555339697967992321677559773500258505
hash for frame 11: 330474274430304744544610948707394382901
hash for frame 12: 163366299378624979519226871422138909135
hash for frame 13: 328848768160161808261791477075051645739
hash for frame 14: 234489622952974924107308950178154033028
hash for frame 15: 131808828857208966393865602076614840631
hash for frame 16: 288663956745888946135832388109013584972
```

Then a delete a line that should be in frame 5

```
hash for frame 0: 166703367601710688816886167189174114022
hash for frame 1: 98367643144455113841213878673356476442
hash for frame 2: 32996071146373462600790487505688723979
hash for frame 3: 250643854608231642061781652711835403862
hash for frame 4: 40494954945001763596001312965364155748
hash for frame 5: 88365673818236020267713969410365157643
hash for frame 6: 335696108594705439680628222702518971520
hash for frame 7: 36488333606598884934589573443182311698
hash for frame 8: 287675856955751747433310497777152853981
hash for frame 9: 202991042687576132822169856435203009984
hash for frame 10: 238112859510469515938526045115974869096
hash for frame 11: 182731103645534480596732477822321140480
hash for frame 12: 321543235758578969515099867632916853400
hash for frame 13: 126676476882648521761792648570683348801
hash for frame 14: 98339766842292100342210093244990071805
hash for frame 15: 131808828857208966393865602076614840631
hash for frame 16: 280184380024170098810519470200135964755
```

You can see that most frames becomes different. Therefore memorizing this doesn't make sense.

Reply:
Maybe we should develop some util to investigate on these "mis-cache" bugs. This should improve the observability of typst, given that we already have tracing.

Reply:
For those who are interested, I'm playing on this line: https://github.com/werifu/HUST-typst-template/blob/main/sample.typ#L32. You can have a try of yourself if you like.

Reply:
Might be because of the spans that it's different. Not that it matters anymore, but `Arc<str>` is basically equivalent to `EcoString` and would have also required a copy when converting from `String` whereas `Arc<String>` wouldn't have.

Reply:
thanks for figure it out. I currently remove memorization for this function and all memorized function now use `EcoString`.